dds/idl/marshal_generator.cpp

```diff
@@ -909,6 +909,422 @@ namespace {
+      AST_Type* const act = AstTypeClassification::resolveActualType(map);
+      AST_Map * const m = dynamic_cast<AST_Map*>(act);
+      key_cxx_elem = scoped(m->key_type()->name());
+      val_cxx_elem = scoped(m->value_type()->name());
+    } else {
+      key_cxx_elem = scoped(key->name());
+      val_cxx_elem = scoped(val->name());
+    }
+
+    RefWrapper (base_wrapper).done().generate_tag();
+
+    {
+      Intro intro;
+      RefWrapper wrapper(base_wrapper);
+      wrapper.done(&intro);
+      const std::string value_access = wrapper.value_access();
+      const std::string get_length = wrapper.map_get_length();
+      const std::string check_empty = wrapper.map_check_empty();
+      Function serialized_size("serialized_size", "void");
+      serialized_size.addArg("encoding", "const Encoding&");
+      serialized_size.addArg("size", "size_t&");
+      serialized_size.addArg("map", wrapper.wrapped_type_name());
+      serialized_size.endArgs();
+
+      if ((key_cls & CL_INTERFACE) == 0 || (val_cls & CL_INTERFACE) == 0) {
+        marshal_generator::generate_dheader_code("    serialized_size_delimiter(encoding, size);\n", !val_primitive, false);
+        intro.join(be_global->impl_, "  ");
+        be_global->impl_ <<
+          "  primitive_serialized_size_ulong(encoding, size);\n"
+          "  if (" << check_empty << ") {\n"
+          "    return;\n"
+          "  }\n";
+      }
+
+      if (key_cls & CL_INTERFACE || val_cls & CL_INTERFACE) {
+        be_global->impl_ <<
+          "  // map with either a key or value of objrefs is not marshaled\n";
+      } else if (key_cls & CL_UNKNOWN || val_cls & CL_UNKNOWN) {
+        be_global->impl_ <<
+          "  // map with either key or value of unknown/unsupported type\n";
+      } else {
+
+        // Simple Types
+        bool key_generated = false;
+        if (key_cls & CL_ENUM) {
+          be_global->impl_ <<
+            "  primitive_serialized_size_ulong(encoding, size, " + get_length + ");\n";
+          key_generated = true;
+        } else if (key_cls & CL_PRIMITIVE) {
+          be_global->impl_ << checkAlignment(key) <<
+            "  " + getSizeExprPrimitive(key, get_length) << ";\n";
+          key_generated = true;
+        }
+
+        bool val_generated = false;
+        if (val_cls & CL_ENUM) {
+          be_global->impl_ <<
+            "  primitive_serialized_size_ulong(encoding, size, " + get_length + ");\n";
+          val_generated = true;
+        } else if (val_cls & CL_PRIMITIVE) {
+          be_global->impl_ << checkAlignment(val) <<
+            "  " + getSizeExprPrimitive(val, get_length) << ";\n";
+          val_generated = true;
+        }
+
+        // Complex Types
+        if (!key_generated || !val_generated) {
+          be_global->impl_ <<
+            "  for (auto it = " << value_access << ".begin(); it != " << value_access << ".end(); ++it) {\n";
+          const std::string indent = "    ";
+
+          if (!key_generated) { // Key
```

Comment:
Would it work to refactor the "Key" and "Value" sections to be calls to a common function?